centralstore/smartwrapper.go

```diff
@@ -15,58 +15,59 @@ type statusMap map[string]*CentralTraceStatus
 
 // This is an implementation of SmartStorer that stores spans in memory locally.
 type SmartWrapper struct {
-	basicStore     BasicStorer
+	Config         config.Config `inject:""`
+	BasicStore     BasicStorer   `inject:""`
+	Tracer         trace.Tracer  `inject:"tracer"`
 	keyfields      []string
-	stopped        chan struct{}
+	closing        bool
 	spanChan       chan *CentralSpan
+	stopped        chan struct{}
 	done           chan struct{}
 	doneProcessing chan struct{}
-	tracer         trace.Tracer
 }
 
 // ensure that we implement SmartStorer
 var _ SmartStorer = (*SmartWrapper)(nil)
 
-// this probably gets moved into config eventually
-type SmartWrapperOptions struct {
-	SpanChannelSize int
-	StateTicker     config.Duration
-	SendDelay       config.Duration
-	TraceTimeout    config.Duration
-	DecisionTimeout config.Duration
-}
-
-// NewSmartWrapper creates a new SmartWrapper.
-func NewSmartWrapper(options SmartWrapperOptions, basic BasicStorer, tracer trace.Tracer) *SmartWrapper {
-	w := &SmartWrapper{
-		basicStore:     basic,
-		stopped:        make(chan struct{}),
-		spanChan:       make(chan *CentralSpan, options.SpanChannelSize),
-		done:           make(chan struct{}),
-		doneProcessing: make(chan struct{}),
-		tracer:         tracer,
+// Start validates and starts the SmartWrapper.
+func (w *SmartWrapper) Start() error {
+	// check that the injection is doing its job
+	if w.Config == nil {
+		return fmt.Errorf("config is nil")
+	}
+	if w.Tracer == nil {
+		return fmt.Errorf("tracer is nil")
 	}
+	if w.BasicStore == nil {
+		return fmt.Errorf("basicStore is nil")
+	}
+	opts := w.Config.GetCentralStoreOptions()
+
+	w.stopped = make(chan struct{})
+	w.spanChan = make(chan *CentralSpan, opts.SpanChannelSize)
+	w.done = make(chan struct{})
+	w.doneProcessing = make(chan struct{})
+
 	// start the span processor
 	go w.processSpans(context.Background())
 
 	// start the state manager
-	go w.manageStates(context.Background(), options)
+	go w.manageStates(context.Background(), opts)
 
-	return w
+	return nil
 }
 
-func (w *SmartWrapper) Stop() {
+func (w *SmartWrapper) Stop() error {
 	fmt.Println("stopping SmartWrapper")
+	w.closing = true
```

Comment:
Is `closing` being used anywhere? I can't seem to find it

Reply:
It was an experiment that failed; good catch.